linera-core/src/client.rs

```diff
@@ -449,41 +446,71 @@ where
     }
 
     /// Listens to notifications about the current chain from all validators.
-    // TODO(#687): handle reconfigurations, loss of connection, etc.
+    // TODO(#687): handle loss of connection, etc.
     pub async fn listen(this: Arc<Mutex<Self>>) -> Result<NotificationStream>
+    where
+        P: Send + 'static,
+    {
+        let mut streams = HashMap::new();
+        if let Err(err) = Self::update_committee(&this, &mut streams).await {
+            error!("Failed to update committee: {}", err);
+        }
+        let stream = stream! {
+            while let (Some(notification), _, _) =
+                future::select_all(streams.values_mut().map(StreamExt::next)).await
+            {
+                if matches!(notification.reason, Reason::NewBlock { .. }) {
+                    if let Err(err) = Self::update_committee(&this, &mut streams).await {
+                        error!("Failed to update committee: {}", err);
+                    }
+                }
+                yield notification;
+            }
+        };
+        Ok(Box::pin(stream))
+    }
+
+    async fn update_committee(
```

Comment:
IMHO what's updated is the set of streams. Maybe `update_streams[_from_current_committee]` ?